### PR TITLE
Updated asset tracker and pca20035 support

### PR DESCRIPTION
**Pull request by joakimtoe**
Adding:
- Real GPS usage!
- USB UART bridge sample
- Battery monitoring
- Cloud API usage.
- Environment API

### CODE REVIEW

File: drivers/nrf9160_gps/nrf9160_gps.c [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+		}
+
+		print_satellite_stats(&raw_gps_data);
+
+		switch (raw_gps_data.data_id) {
+		case NRF_GNSS_PVT_DATA_ID:
+			copy_pvt(&fresh_pvt.pvt, &raw_gps_data.pvt);
+
+			if ((drv_data->trigger.chan == GPS_CHAN_PVT) &&
+			    (drv_data->trigger.type == GPS_TRIG_DATA_READY)) {
```
Comment (Rallare) [inline drivers/nrf9160_gps/nrf9160_gps.c]: As this module is today, it does not seem to be possible to set the trigger to _DATA_READY and _FIX, so you cannot get both in the application. Wouldn't it be better to send the event to the application regardless of which event is generated? Or expanding the API to accept more than one trigger?

Comment (jtguggedal) [reply drivers/nrf9160_gps/nrf9160_gps.c]: Indeed, this driver and API in general should be reworked, at least if it's still on the "work-in-progress state" where I left it before going on leave.
It's pretty much a copy of the sensor API, with the same limitations regarding triggers, and I've been awaiting the upstream rework of that API to see how GPS would better fit in there.

The plan was to have callback on multiple events, drop the mimicking of the sensor API and hopefully make use of the new upstream work. So if anyone wants to continue that work before I'm back, that would be great.

Comment (Rallare) [reply drivers/nrf9160_gps/nrf9160_gps.c]: What I did as a test was to butcher your implementation, and set the drv_data->trigger.type to the corresponding event, and call the handler() for each corresponding event. This way, I can filter in the application callback on the different events. The init function just omits the trigger type, so not nearly polished :)

File: applications/asset_tracker/CMakeLists.txt [hunk 2]
```diff
@@ -22,9 +22,37 @@ zephyr_library_include_directories(
...
+  src/battery_monitor
   )
 
 # Application sources
 add_subdirectory(src/orientation_detector)
 add_subdirectory(src/ui)
 add_subdirectory(src/cloud_codec)
+add_subdirectory(src/gps_controller)
+if (CONFIG_BATTERY_MONITOR)
+add_subdirectory(src/battery_monitor)
```
Comment (SebastianBoe) [inline applications/asset_tracker/CMakeLists.txt]: indentation

File: applications/asset_tracker/CMakeLists.txt [hunk 3]
```diff
@@ -22,9 +22,37 @@ zephyr_library_include_directories(
...
+endif()
+
+include_directories(src/env_sensors)
+
+if (CONFIG_USE_BME680_BSEC)
+	target_sources(app PRIVATE src/env_sensors//bsec.c)
```
Comment (SebastianBoe) [inline applications/asset_tracker/CMakeLists.txt]: The indentation here looks off, please compare with zephyr/CMakeLists.txt

File: applications/asset_tracker/CMakeLists.txt [hunk 4]
```diff
@@ -22,9 +22,37 @@ zephyr_library_include_directories(
...
 # Application sources
 add_subdirectory(src/orientation_detector)
 add_subdirectory(src/ui)
 add_subdirectory(src/cloud_codec)
+add_subdirectory(src/gps_controller)
+if (CONFIG_BATTERY_MONITOR)
+add_subdirectory(src/battery_monitor)
+endif()
+
+include_directories(src/env_sensors)
```
Comment (SebastianBoe) [inline applications/asset_tracker/CMakeLists.txt]: s/include_directories/zephyr_include_directories/

Comment (SebastianBoe) [reply applications/asset_tracker/CMakeLists.txt]: Even better would be to modify line 18, zephyr_include_directories(src), to be

zephyr_include_directories(
  src
  src/env_sensors
)

or possibly have whatever is trying to include things in env_sensors do #include <src/env_sensors/something_in_env_sensors.h>

File: applications/asset_tracker/CMakeLists.txt [hunk 5]
```diff
@@ -22,9 +22,37 @@ zephyr_library_include_directories(
...
+
+if (CONFIG_USE_BME680_BSEC)
+	target_sources(app PRIVATE src/env_sensors//bsec.c)
+
+	set (bsec_dir ${CMAKE_CURRENT_SOURCE_DIR}/src/env_sensors/BSEC_1.4.7.3_Generic_Release)
```
Comment (SebastianBoe) [inline applications/asset_tracker/CMakeLists.txt]: Put the version in a variable

File: applications/asset_tracker/CMakeLists.txt [hunk 6]
```diff
@@ -22,9 +22,37 @@ zephyr_library_include_directories(
...
+	include_directories(${bsec_dir}/API)
+	target_sources(app PRIVATE ${bsec_dir}/examples/bsec_integration.c)
+	include_directories(${bsec_dir}/examples)
+	target_sources(app PRIVATE ${bsec_dir}/API/bme680.c)
+
+	set(BSEC_LIB_DIR ${bsec_dir}/algo/bin/Normal_version/gcc/Cortex_M4F)
+	set(BSEC_INCLUDE_DIR ${bsec_dir}/algo/bin/Normal_version/gcc/Cortex_M4F)
+	add_library(bsec_lib STATIC IMPORTED GLOBAL)
+	add_dependencies(bsec_lib math_lib bsec_target)
+	set_target_properties(bsec_lib PROPERTIES IMPORTED_LOCATION "${BSEC_LIB_DIR}/libalgobsec.a")
```
Comment (SebastianBoe) [inline applications/asset_tracker/CMakeLists.txt]: You need to enforce somehow that the ABI of libalgobsec.a is compatible with the current build configuration.

See https://github.com/NordicPlayground/nrfxlib/blob/master/ble_controller/CMakeLists.txt#L14 for the current best-practice on how to do this.

File: applications/asset_tracker/CMakeLists.txt [hunk 7]
```diff
@@ -22,9 +22,37 @@ zephyr_library_include_directories(
...
+	target_sources(app PRIVATE src/env_sensors//bsec.c)
+
+	set (bsec_dir ${CMAKE_CURRENT_SOURCE_DIR}/src/env_sensors/BSEC_1.4.7.3_Generic_Release)
+	include_directories(${bsec_dir}/API)
+	target_sources(app PRIVATE ${bsec_dir}/examples/bsec_integration.c)
+	include_directories(${bsec_dir}/examples)
+	target_sources(app PRIVATE ${bsec_dir}/API/bme680.c)
+
+	set(BSEC_LIB_DIR ${bsec_dir}/algo/bin/Normal_version/gcc/Cortex_M4F)
+	set(BSEC_INCLUDE_DIR ${bsec_dir}/algo/bin/Normal_version/gcc/Cortex_M4F)
```
Comment (SebastianBoe) [inline applications/asset_tracker/CMakeLists.txt]: Align these two paths so it is clear they are identical.

File: applications/asset_tracker/CMakeLists.txt [hunk 8]
```diff
@@ -22,9 +22,37 @@ zephyr_library_include_directories(
...
+	add_library(bsec_lib STATIC IMPORTED GLOBAL)
+	add_dependencies(bsec_lib math_lib bsec_target)
+	set_target_properties(bsec_lib PROPERTIES IMPORTED_LOCATION "${BSEC_LIB_DIR}/libalgobsec.a")
+	set_target_properties(bsec_lib PROPERTIES INTERFACE_INCLUDE_DIRECTORIES ${BSEC_INCLUDE_DIR})
+	target_link_libraries(bsec_lib INTERFACE -L${LIBC_LIBRARY_DIR})
+	target_link_libraries(app PUBLIC bsec_lib)
+else()
+	target_sources(app PRIVATE src/env_sensors/env_sensors.c)
```
Comment (SebastianBoe) [inline applications/asset_tracker/CMakeLists.txt]: I believe we should surround this code with "NORDIC SDK APP" markers.

https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/gs_modifying.html#requirements-for-cmakelists-txt

Perhaps @tejlmand can confirm that the parser can deal with complex situations like this.

File: applications/asset_tracker/prj.conf [hunk 9]
```diff
@@ -48,12 +75,24 @@ CONFIG_MAIN_THREAD_PRIORITY=7
...
 
 # Disable native network stack to save some memory
-CONFIG_NET_IPV4=n
-CONFIG_NET_IPV6=n
-CONFIG_NET_UDP=n
-CONFIG_NET_TCP=n
+#CONFIG_NET_IPV4=n
+#CONFIG_NET_IPV6=n
+#CONFIG_NET_UDP=n
+#CONFIG_NET_TCP=n
```
Comment (SebastianBoe) [inline applications/asset_tracker/prj.conf]: There should be a comment that let's the reader know that these options weren't commented out by accident.

File: applications/asset_tracker/src/ui/Kconfig [hunk 10]
```diff
@@ -92,4 +92,4 @@ endmenu
 
 module = UI
 module-str = User Interface
-source "${ZEPHYR_BASE}/subsys/logging/Kconfig.template.log_config"
+source "${ZEPHYR_BASE}/subsys/logging/Kconfig.template.log_config"
```
Comment (SebastianBoe) [inline applications/asset_tracker/src/ui/Kconfig]: ?

missing newline.

File: applications/asset_tracker/src/ui/nmos.c [hunk 11]
```diff
@@ -14,20 +14,18 @@
...
+		NMOS_CONFIG_PIN(_pin), .mode = NMOS_MODE_GPIO                  \
+	}
 
 struct nmos_config {
 	int pin;
-	enum {
-		NMOS_MODE_GPIO,
-		NMOS_MODE_PWM
-	} mode;
+	enum { NMOS_MODE_GPIO, NMOS_MODE_PWM } mode;
```
Comment (SebastianBoe) [inline applications/asset_tracker/src/ui/nmos.c]: These (presumably machine-generated) patches are making the code less readable, so unless we are violating the coding convention they should be reverted. This applies elsewhere as well.

**Comment by hmhalvorsen** [comment]
Note: the commits added here where I am an author are already merged to master or a part of a different PR. They are not required to be a part of this PR.

File: drivers/nrf9160_gps/nrf9160_gps.c [hunk 12]
```diff
@@ -0,0 +1,525 @@
...
+	struct device *dev = INT_TO_POINTER(dev_ptr);
+	struct gps_drv_data *drv_data = dev->driver_data;
+	int len;
+	nrf_gnss_data_frame_t raw_gps_data;
+	bool trigger_send = false;
+wait:
+	k_sem_take(&drv_data->thread_run_sem, K_FOREVER);
+
+	while (true) {
+		len = recv(drv_data->socket, &raw_gps_data,
```
Comment (crfosse) [inline drivers/nrf9160_gps/nrf9160_gps.c]:
```suggestion
		len = nrf_recv(drv_data->socket, &raw_gps_data,
```

I've been having problems with the driver hanging/stopping at this **recv**. From what I can find on Devzone it seems like using **nrf_socket** and **net/socket** "mixed" can lead to undefined behavior as **net/socket** isn't fully compatible with the nRF9160 yet.

In this case the GPS socket is opened with **nrf_socket**, hence **nrf_recv** should be used later on. Did some testing with this and it seemed more stable.

**Comment by thst-nordic** [comment]
In order to run the correct CI you need to update from master to get latest jenkinsfile.

**Pull request closed without merging**